_posts/2018-12-06-extended-memcpy.md

```diff
@@ -0,0 +1,153 @@
+APIs. For libpmem functions we introduced these flags:
+  - `PMEM_F_MEM_NODRAIN`
+  - `PMEM_F_MEM_NOFLUSH`
+  - `PMEM_F_MEM_NONTEMPORAL`
+  - `PMEM_F_MEM_TEMPORAL`
+  - `PMEM_F_MEM_WC`
+  - `PMEM_F_MEM_WB`
+
+`PMEM_F_MEM_NODRAIN` disables `pmem_drain`, just like `_nodrain` functions
+did. 
+
+`PMEM_F_MEM_NOFLUSH` disables flushing (and implies `PMEM_F_MEM_NODRAIN`).
+This flag may be useful if application knows it will update the same piece of
+memory multiple times, doesn't care about possible cache eviction in between
+and still wants to rely on the 8-byte atomicity guarantee.
+
+`PMEM_F_MEM_NONTEMPORAL` tells the library that data will not be used again
+soon (so it can bypass CPU cache). On x86_64 this means the usage of NT-stores.
+
+`PMEM_F_MEM_WC` tells the library to treat destination memory as
+write-combining (so it can avoid CPU cache miss on read). On x86_64 this
+means the usage of NT-stores.
+
+`PMEM_F_MEM_TEMPORAL` tells the library that data might be used again soon
+(so it should keep it in CPU cache). On x86_64 this means the usage of normal
+stores. On x86_64 without CLWB instruction (or motherboard without eADR)
+the cache will still be invalidated (unless `PMEM_F_MEM_NOFLUSH` was also
+used).
+
+`PMEM_F_MEM_WB` tells the library to treat destination memory as write-back.
+On x86_64 this means the usage of normal stores.
+
+Few more notes on non-temporal stores usage on x86_64:
+- If you use `PMEM_F_MEM_NODRAIN`, you should call `pmem_drain` at least
+  when you want the copied data to be visible to other threads, because without
+  that your data may be stuck in (per CPU) write-combining buffers.
+- It's important to align modified data to cache-line boundary (64 bytes),
+  because write-combining happens on cache-line level.
+  In pmemobj you can use allocation classes with multiple of 64 bytes alignment
+  to achieve that (by default objects are allocated using only 16 bytes
+  alignment). 
+
+For libpmemobj functions we introduced these flags: 
+  - `PMEMOBJ_F_MEM_NODRAIN`
+  - `PMEMOBJ_F_MEM_NOFLUSH`
+  - `PMEMOBJ_F_MEM_NONTEMPORAL`
+  - `PMEMOBJ_F_MEM_TEMPORAL`
+  - `PMEMOBJ_F_MEM_WC`
+  - `PMEMOBJ_F_MEM_WB`
+  - `PMEMOBJ_F_RELAXED`
+
+First 6 behave like their libpmem equivalents.
+
+`PMEMOBJ_F_RELAXED` matters only when used with remote replication.
+For now RDMA protocol (as used by librpmem to implement remote replication)
+doesn't guarantee that in case of interruption the transfer won't be torn
+at random place. In order to provide 8 byte atomicity guarantee mentioned
+earlier, pmemobj have to use slower method of replication. This flag tells
+the library that the application doesn't care about this guarantee for
+this operation, so it can replicate using faster method.
+
+If you are overwhelmed by this knowledge, sticking to older functions or setting
+flags argument to 0 for the new ones is a safe choice. However if you want
+to squeeze as much of performance from your brand new NVDIMMs as possible,
```

Comment:
squeeze as much performance as possible from your brand new NVDIMMs, it's may be worth using those new APIs.